File: lab4/cpp/src/RouterNode.cpp
#include "RouterNode.h"
#include "RouterPacket.h"

#include <iomanip>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

/*
 * Initialize the RouterNode.
 * Also synchronize the base costs on the network before the first LINK_CHANGE
 * event by using the notifyNetwork method.
 */
RouterNode::RouterNode(int ID, RouterSimulator* sim, vector<int> const& costs)
    : myGUI{ "  Output window for router #" + to_string(ID) + "  " },
      sim{ sim }, myID{ ID }, costs{ costs },
      distances(sim->NUM_NODES, vector<int>(sim->NUM_NODES, sim->INFINITY)),
      routes(sim->NUM_NODES, "-") {

    // Init distances and routes info.
    distances[myID] = costs;
    for (int i = 0; i < sim->NUM_NODES; i++) {
        if (costs[i] != sim->INFINITY) {
            routes[i] = to_string(i);
        }
    }

    // Notify the network about the starting costs of this node
    notifyNetwork();
}

/*
 * Recalculate the distance costs to all destinations for this node by using:
 * min{ cost(x -> y) + distance(y -> destination) }
 * where x is ourselves and y is any adjacent neighbor.
 */
void RouterNode::updateDistanceCosts() {
    for (int target = 0; target < sim->NUM_NODES; target++) {
        if (target == myID) {
            continue;
        }
        // Calculate all possible routes so that we can find the cheapest one.
        // For each cost: save the first hop at the same index.
        vector<int> targetRoutes;
        vector<int> targetFirstHops;
        targetRoutes.reserve(sim->NUM_NODES);
        targetFirstHops.reserve(sim->NUM_NODES);
        for (int next = 0; next < sim->NUM_NODES; next++) {
            if (next == myID) {
                continue;
            }
            targetRoutes.push_back(costs[next] + distances[next][target]);
            targetFirstHops.push_back(next);
        }
        // Find the minimum cost possible, and its corresponding first hop.
        int minCost = sim->INFINITY;
        int minFirstHopID;
        for (size_t i = 0; i < targetRoutes.size(); i++) {
            if (targetRoutes[i] < minCost) {
                minCost = targetRoutes[i];
                minFirstHopID = targetFirstHops[i];
            }
        }
        this->distances[myID][target] = minCost;
        // Set the route info for informative printing
        this->routes[target] = to_string(minFirstHopID);
    }
}

/*
 * Send our distances to the entire network
 * If POISONREVERSE is true then we should send INFINITY to any node that we are
 * connected to.
 * `fakeindex` defaults to a null pointer.
 */
void RouterNode::notifyNetwork(int* fakeidx) {
    for (int target = 0; target < sim->NUM_NODES; target++) {
        // Prepare a vector that we might need to add poisoned data to
        vector<int> sendvector = distances[myID];
        if (sim->POISONREVERSE && fakeidx != nullptr && target != *fakeidx) {
            sendvector[*fakeidx] = sim->INFINITY;
        }
        if (target == myID) {
            continue;
        }
        if (costs[target] == sim->INFINITY) {
            continue;
        }
        RouterPacket pkt{
            myID,
            target,
            std::move(sendvector),
        };
        sendUpdate(pkt);
    }
}

/*
 * When an update is received, update our distance costs and propagate any
 * updated costs, if there is no change, don't propagate.
 */
void RouterNode::recvUpdate(RouterPacket& pkt) {
    distances[pkt.sourceid] = pkt.mincost;
    vector<vector<int>> oldDistances = distances;
    updateDistanceCosts();
    if (oldDistances != distances) {
        // Send a regular update to the network, without poisoning any data.
        notifyNetwork();
    }
}

/*
 * Send a prepared packet to the simulator.
 */
void RouterNode::sendUpdate(RouterPacket& pkt) {
    sim->toLayer2(pkt);
}

/*
 * Format and print state info about this router node.
 */
void RouterNode::printDistanceTable() {
    // Use a string builder to avoid expensive myGUI.print calls
    ostringstream stringBuilder;
    stringBuilder << "Current state for " << myID << " at time " << std::fixed
                  << setprecision(1) << sim->getClockTime() << "\n\n";

    stringBuilder << "Distancetable\n";
    stringBuilder << "    dst |";
    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << setw(5) << i;
    }
    stringBuilder << '\n';

    stringBuilder << "---------";
    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << "-----";
    }
    stringBuilder << '\n';

    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << " nbr" << setw(4) << i << '|';
        for (int j = 0; j < sim->NUM_NODES; j++) {
            stringBuilder << setw(5) << distances[i][j];
        }
        stringBuilder << '\n';
    }
    stringBuilder << '\n';

    stringBuilder << "Our distance vector and routes:\n";
    stringBuilder << "    dst |";
    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << setw(5) << i;
    }
    stringBuilder << '\n';

    stringBuilder << "---------";
    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << "-----";
    }
    stringBuilder << '\n';

    stringBuilder << " cost   |";
    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << setw(5) << distances[myID][i];
    }
    stringBuilder << '\n';

    stringBuilder << " route  |";
    for (int i = 0; i < sim->NUM_NODES; i++) {
        stringBuilder << setw(5) << routes[i];
    }
    stringBuilder << "\n\n";
    myGUI.println(stringBuilder.str());
}

/*
 * Set a new link cost for this node to another node.
 * Since this might cause an update in distance costs when POISONREVERSE is
 * true
 */
void RouterNode::updateLinkCost(int dest, int newcost) {
    costs[dest] = newcost;
    updateDistanceCosts();
    // Pass what node ID to poison data about, if POISONREVERSE is true
    notifyNetwork(&dest);
}

File: lab4/py/RouterNode.py
#!/usr/bin/env python
import F
from RouterPacket import RouterPacket
from GuiTextArea import GuiTextArea
from RouterSimulator import RouterSimulator

from copy import deepcopy
from typing import Optional
from random import random


class RouterNode():
    # Access simulator variables with:
    # self.sim.POISONREVERSE, self.sim.NUM_NODES, etc.

    def __init__(self, ID: int, sim: RouterSimulator, costs: list[int]):
        self.myGUI = GuiTextArea(
            '  Output window for Router #' + str(ID) + '  ',
        )
        self.sim: RouterSimulator = sim
        self.myID: int = ID
        self.costs: list[int] = costs
        self.distances: list[list[int]] = [
            [999] * self.sim.NUM_NODES
            for _ in range(self.sim.NUM_NODES)
        ]
        self.routes: list[str] = ['-'] * self.sim.NUM_NODES

        # Init distances and routes info
        self.distances[self.myID] = deepcopy(costs)
        for i in range(self.sim.NUM_NODES):
            if self.costs[i] != self.sim.INFINITY:
                self.routes[i] = str(i)
        # Notify other nodes on the network about our distances
        self.notifyNeighbors()

    # --------------------------------------------------

    def updateDistanceCosts(self) -> None:
        for target in range(self.sim.NUM_NODES):
            if target == self.myID:
                continue
            nextRoutes: list[int] = list()
            nextHop: list[int] = list()
            for next in range(self.sim.NUM_NODES):
                if next == self.myID:
                    continue
                nextRoutes.append(
                    self.costs[next] + self.distances[next][target],
                )
                nextHop.append(next)
            minCost: int = self.sim.INFINITY
            minHopID: int
            for next in range(len(nextRoutes)):
                if nextRoutes[next] < minCost:
                    minCost = nextRoutes[next]
                    minHopID = nextHop[next]
            self.distances[self.myID][target] = minCost
            self.routes[target] = str(minHopID)

    # --------------------------------------------------

    def notifyNeighbors(self, fakeindex: Optional[int] = None) -> None:
        if not self.sim.POISONREVERSE:
            for neighborID in range(self.sim.NUM_NODES):
                if neighborID == self.myID:
                    continue
                pkt = RouterPacket(
                    self.myID,
                    neighborID,
                    deepcopy(self.distances[self.myID]),
                )
                self.sendUpdate(pkt)
        else:
            fakevector: list[int] = deepcopy(self.distances[self.myID])
            if fakeindex is not None:
                fakevector[fakeindex] = self.sim.INFINITY
            for neighborID in range(self.sim.NUM_NODES):
                if neighborID == self.myID:
                    continue
                if self.costs[neighborID] == self.sim.INFINITY:
                    continue
                pkt = RouterPacket(
                    self.myID,
                    neighborID,
                    fakevector,
                )
                self.sendUpdate(pkt)

    # --------------------------------------------------

    def recvUpdate(self, node: Optional[RouterPacket]) -> None:
        assert node is not None
        self.distances[node.sourceid] = node.mincost
        oldDistance: list[list[int]] = deepcopy(self.distances)
        self.updateDistanceCosts()
        if oldDistance != self.distances:
            self.notifyNeighbors()

    # --------------------------------------------------

    def sendUpdate(self, node: RouterPacket) -> None:
        assert node is not None
        self.sim.toLayer2(node)

    # --------------------------------------------------

    def printDistanceTable(self) -> None:
        self.myGUI.println(
            f'Current state for router {self.myID} at time ' +
            f'{self.sim.getClocktime():.1f}\n',
        )

        self.myGUI.println('Distancetable')
        self.myGUI.print('    dst |')
        for i in range(self.sim.NUM_NODES):
            self.myGUI.print(f'{i: >5}')
        self.myGUI.println()

        self.myGUI.println('---------' + '-----' * self.sim.NUM_NODES)

        for i in range(self.sim.NUM_NODES):
            self.myGUI.print(f' nbr{i: <4}|')
            for j in range(self.sim.NUM_NODES):
                self.myGUI.print(f'{self.distances[i][j]: >5}')
            self.myGUI.println()
        self.myGUI.println()

        self.myGUI.println('Our distance vector and routes:')
        self.myGUI.print('    dst |')
        for i in range(self.sim.NUM_NODES):
            self.myGUI.print(f'{i: >5}')
        self.myGUI.println()

        self.myGUI.println('---------' + '-----' * self.sim.NUM_NODES)

        self.myGUI.print(' cost   |')
        for i in range(self.sim.NUM_NODES):
            self.myGUI.print(f'{self.distances[self.myID][i]: >5}')
        self.myGUI.println()

        self.myGUI.print(' route  |')
        for i in range(self.sim.NUM_NODES):
            self.myGUI.print(f'{self.routes[i]: >5}')
        self.myGUI.println('\n\n')

    # --------------------------------------------------

    def updateLinkCost(self, dest: int, newcost: int) -> None:
        self.costs[dest] = newcost
        self.updateDistanceCosts()
        self.notifyNeighbors(dest)
